feat: add RPA editor to Camunda Modeler

marstamm opened the pull request:
### Proposed Changes

The source code for the RPA worker can be found here: https://github.com/camunda/rpa-frontend

This PR integrates the RPA Editor UI into the Camunda Modeler.

Core concepts:
 - Connection to the RPA worker is managed by the Editor instance
 - `@camunda/rpa-integration` provides React components for the testing UI, they use the editor instance (required prop) to communicate.

Happy to give sync walkthrough

[screencap-2025-01-23-18-04-17.webm](https://github.com/user-attachments/assets/ccb0abf4-af2a-4af2-beed-87074c21a124)

What will change (UI/Interactions):
 - Links to docs
 - Library Suggestions
 - (potentially) deployment to local instance only


### RPA worker setup

To test against a local worker:
1. Download the [RPA worker from github](https://github.com/camunda/rpa-worker/releases)
2. Add a `rpa-worker.properties` in the same directory as the rpa worker with the content:
    ```
    camunda.rpa.python.extra-requirements=extra-requirements.txt
    camunda.client.zeebe.enabled=false
    ```
2. Add a `extra-requirements.txt` in the same directory as the rpa worker with the content:
    ```
    camunda-utils @ https://github.com/camunda/rpa-python-libraries/releases/download/camunda-utils-v0.1.0/camunda_utils-0.1.0-py3-none-any.whl
    camunda-rpa @ https://github.com/camunda/rpa-python-libraries/releases/download/camunda-rpa-v0.1.0/camunda_rpa-0.1.0-py3-none-any.whl
    ```
   This step will not be required in the 1.0 release
3. Start the RPA worker by running the executable 

### Checklist

To ensure you provided everything we need to look at your PR:

* [x] __Brief textual description__ of the changes present
* [x] __Visual demo__ attached
* [x] __Steps to try out__ present, ==> see comment
* [ ] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--
Thanks for creating this pull request! ❤️
-->

nikku commented:
@marstamm Could you share run instructions for non-developers?

marstamm commented:
However, you can download a build-on-demand build:

- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-linux-x64.tar.gz
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-mac-arm64.dmg
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-mac-arm64.zip
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-mac-x64.dmg
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-mac-x64.zip
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-win-ia32.zip
- https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/rpa-integration/camunda-modeler-rpa-integration-win-x64.zip

Note: this is still a WIP, but the integration work is nearly complete. I'll reach out in you Slack channel to get a review for it later.
WE just don't want to merge it yet, as 
 - Documentation links
 - Library Suggestions

Are not final. These changes will be solved by bumping the RPA library and won't require changes to the integration code

barmac reviewed (commented):
I am not able to give a full review right now, but let's sync on this later on.

barmac commented:
Suggestion: Use a single field for the URL

![image](https://github.com/user-attachments/assets/51e0aaef-9c6c-45be-8f09-e655b8d7e201)

barmac commented:
Suggestion: scroll only to the bottom

![image](https://github.com/user-attachments/assets/633134f8-cb61-4006-a918-dd392f6bec69)

barmac commented:
Question: Can we use Codemirror like we do everywhere in the app?

barmac commented:
To fix: Scrolling

https://github.com/user-attachments/assets/32447c31-c271-4929-a1d4-efd0514cd921

barmac commented on client/src/app/tabs/rpa/__tests__/RPAEditorSpec.js:
```diff
@@ -0,0 +1,264 @@
+
+  describe('#render', function() {
+
+    it('should render with NO xml', function() {
```
JSON?

barmac commented:
@marstamm to add a file which we can run with the docker worker.
Also we need a macos image.

barmac commented:
@philippfromme @marstamm and I looked at the PR, and tested the solution in action. View our remarks above.
@marstamm's introduction was recorded and can be accessed at https://camunda.slack.com/archives/C032H77434N/p1738685029197899?thread_ts=1737736287.487929&cid=C032H77434N

marstamm commented:
> To fix: Scrolling

This is caused by the Iframe URL being unreachable. This will display the run log with the production Worker

marstamm commented:
> This is caused by the Iframe URL being unreachable. This will display the run log with the production Worker

This is now fixed. I added a new getting started guide in the main body that used the latest RPA worker, please try it out :)

nikku commented:
I gave this a quick review. The current integration (testing, deploy button) looks a little bit like an alien to me, it does not follow the Desktop modeler UI.

Did we/do we consider sensible adjustments to make it feel more native, i.e. with theeming (https://github.com/camunda/camunda-modeler/issues/4511)? This is something that the core team could follow-up on.

CC @lmbateman @barmac, let's ensure this integrates nicely enough for the users, or decide how/when we get to a "nicely enough" integration.

nikku commented:
@marstamm Is there a reason this is still in draft? What kind of feedback do you seek before moving it to _ready for review_?

marstamm commented:
We still need to define how we want to ship this before the 8.7.0 release, e.g. release behind a feature flag for the alphas

marstamm commented:
> Did we/do we consider sensible adjustments to make it feel more native, i.e. with theeming (https://github.com/camunda/camunda-modeler/issues/4511)? This is something that the core team could follow-up on.

The Components are designed to be used by both desktop and web modeler, to avoid maintaining multiple implementations. 

The Carbon styles are taken from the parent application. Currently, these are just the default carbon styles, but can be changed if desired:

https://github.com/camunda/camunda-modeler/blob/9470f9f90ae23db7e5b9af2d25d7929fea8d99a1/client/src/app/tabs/rpa/RPAEditor.less#L1

lmbateman commented:
@abdul99ahad This might be your first Carbon-theming project. :) Here are my initial thoughts:

- Empty state
  - Can we hide the graphic in the Desktop Modeler?
  - The font size is 16px; in the Variables tab empty state, the font size is 14px.
  - Can we restyle the button and link to Desktop Modeler standard?
- Footer
  - I’m not sure what the new icon is supposed to be - a test tube? There are a few "test" icons in Carbon, but I'm not sure any of them really convey the idea well (or translate to the Desktop Modeler).
  - The “not connected” icon is the wrong color; I think it’s #ff0000, and should be #cc0000.
- Welcome popup
  - The popup is very wide. Not sure if this is a real issue, but it doesn’t fit with our other popups. Also, typographical principles state that a line should be between 40 and 80 characters long. The longest line is just over 80 characters, so we may want to reduce the width a bit for that reason.
  - **Additional configurations** accordion
    - The accordion is a new component for the Desktop Modeler. I think it's the best component for progressive disclosure, but I wonder if we can just show the Base URL field and state that it's optional, since it's only a single field. The accordion feels a bit like overkill.
    - When it’s focused, the accordion looks like a dropdown; I’m not sure there’s anything we can do here aside from not focusing it, which seems suboptimal.
    - When it’s not focused, the accordion does look a bit alien. If we keep it, can we style it a bit in the Desktop Modeler? Darkening the borders might help; it looks like the current color is #e0e0e0, and the borders in the properties panel are #b9bcc6, so significantly darker.
    - Base URL text field: Can we style it like the Properties panel input fields? I see differences in the border color and width (on three sides); the background color (slightly); and the border radius.

I couldn’t figure out how to progress past this point.

lmbateman commented:
Just to clarify my previous comment: My questions about changes were for the Desktop Modeler team. We can hopefully make all the changes I mentioned using a theme that won't affect the styling in Web Modeler. I don't expect the RPA team to make these changes. cc: @PhilippaC22

marstamm commented:
I think most of the things you mentioned can be fixed with some CSS.

> The accordion feels a bit like overkill.

That is something we should fix in the RPA Component instead of the modeler. I'll see if I have time this week. We can also push UI fixes later, as long as the base functionality is not broken

marstamm commented:
What is your preferred solution for shipping this feature? Do you want to delay merging until the 8.7 release, add it with the next alpha, add feature toggles?

Seems like the design concerns can be either tackled with CSS or be done as a patch later. Is there anything blocking this PR once we decided on a release strategy?

barmac commented:
![image](https://github.com/user-attachments/assets/648ae0a6-d8dd-49bc-a7e3-9f35fbc8e99e)
This link should point straight to /docs, not /docs/next

barmac commented:
![image](https://github.com/user-attachments/assets/021e6162-9b4d-43c7-9d0c-a0d797a7808a)
Probably we want to use Desktop Modeler X

barmac commented:
@marstamm and I have just synced on the PR. There are still parts of the contribution which require improvements. However, this should not be a blocker for the entire feature. Since the RPA on the engine is still an alpha feature, we agreed to ship the Modeler part behind a feature flag `--disable-rpa=true`. This will give us space for targeted improvements without blocking the current pull request. We will also gather initial feedback from the users who decide to enable the editor.
Then, for the stable release, we will only need to flip the default value of the flag.

marstamm commented:
I added the discussed changes, you now need to start the modeler with `--no-disable-rpa` flag to see the new Feature

barmac commented:
I was able to get this running locally:
![image](https://github.com/user-attachments/assets/cbee8eaf-4363-4ba7-b349-a95324583fe7)

barmac commented:
@marstamm I think we still want to prevent navigation link from appearing in the report, or at least open them externally (`target=_blank`). Nevertheless, I don't think this can do any harm in Desktop Modeler.

marstamm commented:
> @marstamm I think we still want to prevent navigation link from appearing in the report, or at least open them externally (target=_blank). Nevertheless, I don't think this can do any harm in Desktop Modeler.

I'll see what we can do as the report is generated by RobotFramework and seems to use some sort of JS framework for rendering
